### PR TITLE
Feature/tbd 18/di improvement

### PR DESCRIPTION
**Pull request by SergiiTao**
depends on oat-sa/generis#766

in order to improve DI and get read of `$this->getServiceLocator()->get()` from classes
a class can be compiled into a factory which can use the service locator

```php
// TmpA extends InjectionAwareService
// TmpB extends InjectionAwareService
// TmpD extends ConfigurationService and does not have SERVICE_ID const
// TmpC extends ConfigurationService and does not have SERVICE_ID const
// TmpE extends DefaultUrlService extends ConfigurationService has SERVICE_ID


$this->getServiceManager()->register(
    TmpA::SERVICE_ID,
    new TmpA(
        new TmpB(
            new TmpD()
        ),
        new TmpC(),
        new TmpE()
    )
);
```
it will be automatically compiled  into 
```php
return new class implements \oat\oatbox\service\ServiceFactoryInterface {
    public function __invoke(\Zend\ServiceManager\ServiceLocatorInterface $serviceLocator)
    {
        return new oat\tao\model\service\TmpA(
            new oat\tao\model\service\TmpB(
                $serviceLocator->get(oat\tao\model\service\TmpD::class)
            ),
            $serviceLocator->get(oat\tao\model\service\TmpC::class),
            $serviceLocator->get(oat\tao\model\service\TmpE::SERVICE_ID)
        );
    }
};
```

### CODE REVIEW

File: models/classes/service/InjectionAwareService.php [hunk 1]
```diff
@@ -28,11 +28,25 @@
...
     {
+        $content = 'new %s(%s)';
+
+        if (!$this->isChildItem && $this->isFactory()) {
+            $content = "new class implements \\oat\\oatbox\\service\\ServiceFactoryInterface {
+    public function __invoke(\\Zend\\ServiceManager\\ServiceLocatorInterface \$serviceLocator)
+    {
+        return new %s(%s);
+    }
+}";
```
Comment (wazelin) [inline models/classes/service/InjectionAwareService.php:45]:
```suggestion
            $content = <<<'PHP'
new class implements \oat\oatbox\service\ServiceFactoryInterface {
    public function __invoke(\Zend\ServiceManager\ServiceLocatorInterface $serviceLocator)
    {
        return new %s(%s);
    }
}
PHP;
```

File: models/classes/service/InjectionAwareService.php [hunk 2]
```diff
@@ -77,9 +89,52 @@ protected function getDependencies(): array
...
+        $dependencies = [];
+
+        foreach ($this->iterateParameters() as $parameter) {
+            $propertyValue = $parameter;
+            if (is_object($propertyValue)) {
+                if (($propertyValue instanceof self)) {
+                    $propertyValue->isChildItem = true;
+                } elseif ($propertyValue instanceof ConfigurableService) {
+                    $className = get_class($propertyValue);
+                    $value = defined("$className::SERVICE_ID") ? "$className::SERVICE_ID" : "$className::class";
```
Comment (wazelin) [inline models/classes/service/InjectionAwareService.php:111]:
```suggestion
                    $serviceIdentifier = defined("$className::SERVICE_ID") ? "$className::SERVICE_ID" : "$className::class";
```

File: models/classes/service/InjectionAwareService.php [hunk 3]
```diff
@@ -77,9 +89,52 @@ protected function getDependencies(): array
...
+            $dependencies[] = $propertyValue;
         }
 
         return $dependencies;
     }
+
+    /**
+     * @throws ReflectionException
+     */
+    protected function isFactory(): bool
```
Comment (wazelin) [inline models/classes/service/InjectionAwareService.php:126]: Would you be so kind to elaborate on that method a bit more, like in a comment? So far I don't really understand what it represents.

File: models/classes/service/InjectionAwareService.php [hunk 4]
```diff
@@ -77,9 +91,52 @@ protected function getDependencies(): array
...
-            $dependencies[] = $classProperty->getValue($this);
+            yield $classProperty->getValue($this);
+        }
+    }
+
+    /**
+     * @return array A list of dependencies to be injected in their order.
+     * @throws ReflectionException
+     */
+    protected function getDependencies(): array
```
Comment (wazelin) [inline models/classes/service/InjectionAwareService.php:102]: This desperately needs an integration test.
Like a service stub, implementing that `InjectionAwareService`, and test cases on top of it.

Comment (wazelin) [reply models/classes/service/InjectionAwareService.php:102]: Other than that, good job! Impressive.

Comment (SergiiTao) [reply models/classes/service/InjectionAwareService.php:102]: added

File: models/classes/service/InjectionAwareService.php [hunk 5]
```diff
@@ -77,9 +93,61 @@ protected function getDependencies(): array
...
+            if (
+                is_object($propertyValue)
+                && $propertyValue instanceof ConfigurableService
+            ) {
+                if (!($propertyValue instanceof self)) {
+                    return true;
+                }
+
+                $result = $this->isFactoryNeeded($propertyValue);
+                if ($result) {
```
Comment (wazelin) [inline models/classes/service/InjectionAwareService.php:145]:
```suggestion
            if (!$propertyValue instanceof ConfigurableService) {
                continue;
            }

            if (!$propertyValue instanceof self) {
                return true;
            }

            if ($this->isFactoryNeeded($propertyValue)) {
```

**Pull request merged**